runtime/gc_base/IndexableObjectAllocationModel.cpp

```diff
@@ -133,30 +160,44 @@ MM_IndexableObjectAllocationModel::initializeIndexableObject(MM_EnvironmentBase
 #endif /* defined(J9VM_ENV_DATA64) */
 		} else {
 			indexableObjectModel->setSizeInElementsForContiguous(spine, _numberOfIndexedFields);
+			isDataAdjacentToHeader = indexableObjectModel->isDataAdjacentToHeader(spine);
+			if (isDataAdjacentToHeader) {
 #if defined(J9VM_ENV_DATA64)
-			if (((J9JavaVM *)env->getLanguageVM())->isIndexableDataAddrPresent) {
-				indexableObjectModel->setDataAddrForContiguous(spine);
-			}
+				if (((J9JavaVM *)env->getLanguageVM())->isIndexableDataAddrPresent) {
+					indexableObjectModel->setDataAddrForContiguous(spine);
+				}
 #endif /* defined(J9VM_ENV_DATA64) */
+			} else if (isVirtualLargeObjectHeapEnabled) {
+#if defined(J9VM_ENV_DATA64)
+				/* set NULL temporarily to avoid possible complication with GC occurring while the object is partially initialized? */
+				indexableObjectModel->setDataAddrForContiguous(spine, NULL);
+#endif /* defined(J9VM_ENV_DATA64) */
+			}
 		}
 	}
 
 
 	/* Lay out arraylet and arrayoid pointers */
 	switch (_layout) {
 	case GC_ArrayletObjectModel::InlineContiguous:
-		/**
-		* _numberOfArraylets is set to 0 for 0-stride flattened array, and we
-		* can recognize this case by checking if _dataSize is 0.
-		* Note there are two cases that _dataSize is 0: 0-stride flattened array or 0-length array
-		* But the latter is treated as discontiguous and not following this path
-		*/
-		Assert_MM_true((0 == _dataSize) || (1 == _numberOfArraylets));
+#if defined(J9VM_GC_ENABLE_SPARSE_HEAP_ALLOCATION)
+		if (isVirtualLargeObjectHeapEnabled && !isDataAdjacentToHeader) {
+			/* We still need to create leaves for discontiguous arrays that will be allocated at off-heap */
+			spine = getSparseAddressAndDecommitLeaves(env, spine);
+			if (NULL != spine) {
+				Assert_MM_true(1 <= _numberOfArraylets);
+			}
+		}
+#endif /* defined (J9VM_GC_ENABLE_SPARSE_HEAP_ALLOCATION) */
+		if ((!isVirtualLargeObjectHeapEnabled) || isDataAdjacentToHeader) {
+			Assert_MM_true(1 == _numberOfArraylets);
+		}
 		break;
 
 	case GC_ArrayletObjectModel::Discontiguous:
 	case GC_ArrayletObjectModel::Hybrid:
 		if (NULL != spine) {
+			/* The array size has been set earlier at line 163 */
```

Comment:
please remove "at line 163". Line number is correct now but can be different if code is changed at the future